crates/nri-resctrl-plugin/src/lib.rs

```diff
@@ -20,6 +20,53 @@ use nri::events_mask::EventMask;
 
 use resctrl::{Config as ResctrlConfig, FsProvider, RealFs, Resctrl};
 
+/// Source of PIDs for a container based on cgroup metadata.
+pub trait CgroupPidSource: Send + Sync {
+    fn pids_for_container(&self, c: &nri::api::Container) -> Vec<i32>;
+}
+
+pub struct RealCgroupPidSource;
+
+impl RealCgroupPidSource {
+    pub fn new() -> Self {
+        Self
+    }
+}
+
+#[cfg(target_os = "linux")]
+impl CgroupPidSource for RealCgroupPidSource {
+    fn pids_for_container(&self, c: &nri::api::Container) -> Vec<i32> {
+        // Use cgroups-rs if available by path; fallback to reading files via std fs
+        let cg_path = c
+            .linux
+            .as_ref()
+            .map(|l| l.cgroups_path.clone())
+            .unwrap_or_default();
+        if cg_path.is_empty() {
+            return vec![];
+        }
+        // Try reading cgroup.procs directly
+        let procs_path = std::path::PathBuf::from(&cg_path).join("cgroup.procs");
+        let tasks_path = std::path::PathBuf::from(&cg_path).join("tasks");
+        let s = std::fs::read_to_string(&procs_path)
+            .or_else(|_| std::fs::read_to_string(&tasks_path));
+        match s {
+            Ok(content) => content
+                .lines()
+                .filter_map(|l| l.trim().parse::<i32>().ok())
+                .collect(),
+            Err(_) => vec![],
+        }
+    }
+}
+
+#[cfg(not(target_os = "linux"))]
+impl CgroupPidSource for RealCgroupPidSource {
+    fn pids_for_container(&self, _c: &nri::api::Container) -> Vec<i32> {
+        vec![]
+    }
+}
+
 /// Assignment state for associating pods to resctrl groups.
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub enum AssignmentState {
```

Comment:
now that we have total_containers, reconciled_containers to signify partial or full reconciliation, we can call this ResctrlGroupState where it is either Exists(group_path) or Failed. Then this can replace group_path as well.